javalib/src/main/scala/java/math/BigInteger.scala

```diff
@@ -0,0 +1,1500 @@
+      this
+    } else if (n > 0) {
+      BitLevel.shiftLeft(this, n)
+    } else {
+      BitLevel.shiftRight(this, -n)
+    }
+  }
+
+  /**
+   * Returns a new {@code BigInteger} whose value is {@code this >> n}. For
+   * negative arguments, the result is also negative. The shift distance may be
+   * negative which means that {@code this} is shifted left.
+   * <p>
+   * <b>Implementation Note:</b> Usage of this method on negative values is not
+   * recommended as the current implementation is not efficient.
+   *
+   * @param n shift distance
+   * @return {@code this >> n} if {@code n >= 0}; {@code this << (-n)} otherwise
+   */
+  def shiftRight(n: Int): BigInteger = {
+    if ((n == 0) || (sign == 0)) {
+      this
+    } else if (n > 0) {
+      BitLevel.shiftRight(this, n)
+    } else {
+      BitLevel.shiftLeft(this, -n)
+    }
+  }
+
+  /**
+   * Returns the sign of this {@code BigInteger}.
+   *
+   * @return {@code -1} if {@code this < 0}, {@code 0} if {@code this == 0},
+   *         {@code 1} if {@code this > 0}.
+   */
+  def signum(): Int = sign
+
+  /**
+   * Returns a new {@code BigInteger} whose value is {@code this - val}.
+   *
+   * @param bi value to be subtracted from {@code this}.
+   * @return {@code this - val}.
+   * @throws NullPointerException if {@code val == null}.
+   */
+  def subtract(bi: BigInteger): BigInteger = Elementary.subtract(this, bi)
+
+  /**
+   * Tests whether the bit at position n in {@code this} is set. The result is
+   * equivalent to {@code this & (2^n) != 0}.
+   * <p>
+   * <b>Implementation Note:</b> Usage of this method is not recommended as the
+   * current implementation is not efficient.
+   *
+   * @param n position where the bit in {@code this} has to be inspected.
+   * @return {@code this & (2^n) != 0}.
+   * @throws ArithmeticException if {@code n < 0}.
+   */
+  def testBit(n: Int): Boolean = {
+    var _n = n
+    val intCount = _n >> 5
+
+    if (_n == 0) {
+      (digits(0) & 1) != 0
+    } else if (_n < 0) {
+      throw new ArithmeticException("Negative bit address")
+    } else if (intCount >= numberLength) {
+      sign < 0
+    } else if (sign < 0 && intCount < getFirstNonzeroDigit) {
+      false
+    } else {
+        var digit = digits(intCount)
+        if (sign < 0) {
+            digit = if (getFirstNonzeroDigit == intCount) -digit else ~digit
+        }
+        _n = 1 << (_n & 31)
+        (digit & _n) != 0
+    }
+  }
+
+  /**
+   * Returns the two's complement representation of this BigInteger in a byte
+   * array.
+   *
+   * @return two's complement representation of {@code this}.
+   */
+  /* Todo: This is the original version of the above method, copied here for refence in CR - Needs deleting when CR done
+
+   def toByteArray(): Array[Byte] = {
+   if (this.sign == 0) {
+     return Array[Byte](0)
+   }
+   val temp: BigInteger = this
+   val bitLen = bitLength()
+   val iThis = getFirstNonzeroDigit
+   var bytesLen = (bitLen >> 3) + 1
```

Comment:
It will be gone in next push. This was the largest method I refactored so just put it in as a reference.